sql/wsrep_mysqld.cc

```diff
@@ -2242,6 +2296,216 @@ static void wsrep_TOI_end(THD *thd) {
   }
 }
 
+static int wsrep_NBO_begin(THD *thd, const char *db_, const char *table_,
+                           const TABLE_LIST *table_list,
+                           dd::Tablespace_table_ref_vec *trefs,
+                           Alter_info *alter_info,
+                           wsrep::key_array *fk_tables) {
+  uchar *buf(0);
+  size_t buf_len(0);
+  int buf_err;
+  int rc;
+
+  WSREP_DEBUG("Starting NBO phase one");
+
+  mysql_mutex_lock(&thd->LOCK_thd_data);
+
+  bool is_thd_killed = thd->is_killed();
+
+  mysql_mutex_unlock(&thd->LOCK_thd_data);
+
+  if (is_thd_killed) {
```

Comment:
Not related to this issue. But when the NBO query is killed in between the execution, it seems that the cluster enters into `Disconnected` state.

```
Session 1:
mysql> SET GLOBAL debug = 'd,sync.alter_locked_tables_inplace';
Query OK, 0 rows affected (0.00 sec)

Session 2:
mysql> show processlist;
+----+-----------------+-----------------+------+---------+------+--------------------------+-----------------------------------------+---------+-----------+---------------+
| Id | User            | Host            | db   | Command | Time | State                    | Info                                    | Time_ms | Rows_sent | Rows_examined |
+----+-----------------+-----------------+------+---------+------+--------------------------+-----------------------------------------+---------+-----------+---------------+
|  1 | system user     |                 | NULL | Sleep   | 7086 | wsrep aborter idle       | NULL                                    | 7086340 |         0 |             0 |
|  2 | system user     |                 | NULL | Sleep   | 6193 | innobase_commit_low (-1) | NULL                                    | 6192907 |         0 |             0 |
|  7 | event_scheduler | localhost       | NULL | Daemon  | 7084 | Waiting on empty queue   | NULL                                    | 7084327 |         0 |             0 |
| 18 | root            | localhost:38330 | test | Query   |    6 | debug sync point: now    | alter table t3 add column k varchar(10) |    6345 |         0 |             0 |
| 20 | root            | localhost:55980 | test | Query   |    0 | init                     | show processlist                        |       0 |         0 |             0 |
+----+-----------------+-----------------+------+---------+------+--------------------------+-----------------------------------------+---------+-----------+---------------+
5 rows in set (0.01 sec)

mysql> kill 18;
Query OK, 0 rows affected (0.00 sec)

Session 1:
mysql> alter table t3 add column l varchar(10);
ERROR 1047 (08S01): WSREP has not yet prepared node for application use

mysql> show status like 'wsrep_cluster_status%';
+----------------------+--------------+
| Variable_name        | Value        |
+----------------------+--------------+
| wsrep_cluster_status | Disconnected |
+----------------------+--------------+
1 row in set (0.00 sec)
```